sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -363,7 +363,11 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
             val valueRowBuffer = new Array[Byte](valueSize)
             ByteStreams.readFully(input, valueRowBuffer, 0, valueSize)
             val valueRow = new UnsafeRow(valueSchema.fields.length)
-            valueRow.pointTo(valueRowBuffer, valueSize)
+            // If valueSize in existing file is not multiple of 8, floor it to multiple of 8.
+            // This is work around for the following.
+            // Pre-Spark 2.3 mistakenly append 4 bytes to the value row in
+            // `FixedLengthRowBasedKeyValueBatch`, which gets persisted into the checkpoint data
+            valueRow.pointTo(valueRowBuffer, (valueSize / 8) * 8)
```

Comment:
yes I'm absolutely sure, as we store aggregate buffers in checkpoint files.

BTW, I run [this test](https://github.com/apache/spark/pull/18503/commits/01597010145f7769506cfa51c0450a2db779fc07#diff-b601a70172e88e90cea0020b978d0b5bL482) without the workaround here, and it can still pass. Which means, after this PR, `RowBasedKeyValueBatch` is fixed, and the generated checkpoint files have corrected row length.

And I also run the test without the fix for `RowBasedKeyValueBatch`, and it fails. I think this proves that `RowBasedKeyValueBatch` is the reason why we have wrong row length in checkpoit files.